File: Day2/AdventDay2_part2.py
if (__name__ == '__main__'):
    # Open the input file "input2.txt" in read mode
    with open("input2.txt", "r") as f:
        lines = f.readlines()  # Read all lines from the file into a list

    # Initialize color score variables
    red = blue = green = 0  
    max_red = max_blue = max_green = 0  
    final_sum = 0  # To accumulate the final sum of products
    reset = False  # Flag to determine when to reset scores

    # Process each line in the input file
    for line in lines:
        # Reset color scores and their maximums for each line
        red = blue = green = 0  
        max_red = max_blue = max_green = 0  
        
        # Clean up the line: remove commas and split into words
        words = line.replace(',', '').split()
        
        # Iterate over each word in the processed line
        for i in range(len(words)):
            # Check if the current word ends with a semicolon
            if words[i][len(words[i]) - 1] == ';':
                reset = True  # Mark to reset scores
                words[i] = words[i].replace(';', '')  # Remove the semicolon from the word
            
            # Update the red score if the current word is "red"
            if words[i] == "red":
                red += int(words[i - 1])  # Add the score before "red"
                max_red = max(red, max_red)  # Update maximum red score

            # Update the green score if the current word is "green"
            elif words[i] == "green":
                green = int(words[i - 1])  # Set green score directly from the previous number
                max_green = max(green, max_green)  # Update maximum green score

            # Update the blue score if the current word is "blue"
            elif words[i] == "blue":
                blue += int(words[i - 1])  # Add the score before "blue"
                max_blue = max(blue, max_blue)  # Update maximum blue score

            # If a reset signal was detected, reset all color scores
            if reset:
                red = blue = green = 0  # Reset scores to zero
                reset = False  # Clear reset flag
        
        # Calculate the product of maximum scores and add to the final sum
        final_sum += max_red * max_green * max_blue

    # Output the final accumulated sum of products
    print(final_sum)

File: Day2/AdventDay2_part2.cpp
#include <fstream>  
#include <iostream> 
#include <string>   

using namespace std; 

// Function to find and return a number from a specific starting position in the line
unsigned int find_number(int start, string line) {
    unsigned int j = 0;  // Offset for backtracking from the start position
    int number = 0;      // Variable to hold the found number

    // Continue until a non-digit character is encountered
    while (line[start - j] >= 48 && line[start - j] <= 57) {
        // Convert the substring starting from the current position to an integer
        number = stoi(&line[start - j]);  // Convert string to integer
        j++;  // Move one character to the left
    }
    return number;  // Return the found number
}
  
int main() { 
    // Initialize color score variables
    unsigned int red = 0, green = 0, blue = 0;
    unsigned int max_red = 0, max_green = 0, max_blue = 0;
    int sum = 0;  // To accumulate the final sum of products
    
    // Open the input file for reading
    ifstream inputFile("input2.txt"); 
    if (!inputFile.is_open()) { 
        cerr << "Error opening the file!" << endl;  // Error message if file can't be opened
        return 1; 
    } 
    
    string line;  // Variable to hold each line of the file
    // Read each line from the input file
    while (std::getline(inputFile, line)) {
        // Reset color scores and their maximums for each line
        red = green = blue = 0;
        max_red = max_green = max_blue = 0;

        // Iterate over each character in the line
        for (int i = 0; i < line.size(); i++) {
            // Check for the color "red"
            if (line.substr(i, 3) == "red") {
                // Update the red score based on the preceding number
                red = red + find_number(i - 2, line);
                max_red = max(red, max_red);  // Update maximum red score
            }
            // Check for the color "green"
            else if (line.substr(i, 5) == "green") {
                // Update the green score based on the preceding number
                green = green + find_number(i - 2, line);
                max_green = max(green, max_green);  // Update maximum green score
            }
            // Check for the color "blue"
            else if (line.substr(i, 4) == "blue") {
                // Update the blue score based on the preceding number
                blue = blue + find_number(i - 2, line);
                max_blue = max(blue, max_blue);  // Update maximum blue score
            }
            // If a semicolon is found, reset all color scores
            else if (line[i] == ';') {
                red = 0;   // Reset red score
                green = 0; // Reset green score
                blue = 0;  // Reset blue score
            }
        }

        // Calculate the product of maximum scores and add to the sum
        sum = sum + (max_blue * max_green * max_red);
    }

    // Output the final accumulated sum of products
    cout << sum;  
    return 0;  // Exit program
}
